File: static/js/meridian/payday.js
import { freshnessText, meridianFetch } from "./api.js";
import { formatCurrency } from "./format.js";

const root = document.querySelector("[data-payday-root]");
let currentPayload = null;

function humanDate(value) {
  if (!value) return "—";
  const parsed = new Date(`${value}T12:00:00`);
  if (Number.isNaN(parsed.getTime())) return value;
  return new Intl.DateTimeFormat(undefined, { month: "long", day: "numeric" }).format(parsed);
}

function titleCase(value) {
  return String(value || "").replaceAll("_", " ").replace(/\b\w/g, (letter) => letter.toUpperCase());
}

function isoDay(date) {
  const pad = (value) => String(value).padStart(2, "0");
  return `${date.getFullYear()}-${pad(date.getMonth() + 1)}-${pad(date.getDate())}`;
}

function yesterday() {
  /* The owner asked for the reset to start "at yesterday", so that is the default the
     control offers rather than an unset field. */
  const value = new Date();
  value.setDate(value.getDate() - 1);
  return isoDay(value);
}

function renderLearning(learning) {
  const summary = root.querySelector("[data-learning-floor-summary]");
  const input = root.querySelector("[data-learning-floor-input]");
  const status = root.querySelector("[data-learning-floor-status]");
  if (!summary) return;

  const active = Boolean(learning?.active);
  if (active) {
    summary.textContent =
      `Learning from ${humanDate(learning.floor)} onward · ${learning.included} deposits in use` +
      ` · ${learning.excluded} earlier excluded.`;
  } else {
    summary.textContent = "Learning from all of your income history.";
  }
  // Never clobber what the owner is currently typing.
  if (input && !input.value) input.value = learning?.floor || yesterday();
  if (status && active && learning.set_at) {
    status.textContent = `Set ${new Date(learning.set_at).toLocaleString()}. No records were deleted.`;
  }
}

function cadenceSourceText(cadence) {
  /* Where the cadence came from, said plainly. Before OS-104 the card reported the observed
     pattern alone, so a Crew-held paycheck left it reading "Not recognized" while the same page
     listed that paycheck's cadence further down. When Crew holds more than one record the label
     names the one it used and admits the others: a single figure must not silently stand in for
     several. */
  if (!cadence || !cadence.value) {
    return "No paycheck record in Crew and no deposit pattern yet. Set the paycheck in Crew, or let Meridian learn it from deposits.";
  }
  if (cadence.source === "crew") {
    const named = cadence.plan_name ? ` · ${cadence.plan_name}` : "";
    const others =
      cadence.distinct_cadences > 1
        ? ` (one of ${cadence.distinct_cadences} records)`
        : "";
    return `${cadence.source_label}${named}${others}`;
  }
  return `${cadence.source_label} · ${Math.round((cadence.confidence || 0) * 100)}% confidence · ${
    cadence.deposits || 0
  } deposits`;
}

function render(payload) {
  currentPayload = payload;
  const pattern = payload.pattern;
  const cadence = payload.cadence || { value: null };
  root.querySelector("[data-payday-pattern]").textContent = cadence.value
    ? titleCase(cadence.value)
    : "Not recognized";
  root.querySelector("[data-cadence-source]").textContent = cadenceSourceText(cadence);
  root.querySelector("[data-next-payday]").textContent = humanDate(pattern?.next_date);
  root.querySelector("[data-typical-income]").textContent = pattern
    ? `${formatCurrency(pattern.typical_amount, "USD")} typical income`
    : "Not projected from your deposits yet";

  const nextRun = payload.next_run;
  root.querySelector("[data-next-run-total]").textContent = nextRun
    ? formatCurrency(nextRun.total, "USD")
    : "—";
  root.querySelector("[data-next-run-date]").textContent = nextRun
    ? `${humanDate(nextRun.date)} · proposal only`
    : "Nothing to propose yet";

  const freshness = freshnessText(payload.data_freshness);
  const freshnessNode = root.querySelector("[data-payday-freshness]");
  freshnessNode.dataset.state = freshness.state;
  freshnessNode.textContent = freshness.label;

  /* OS-104 removed this pane's own per-commitment funding editor: it was a two-mode copy of the
     four-mode editor Plan already ships against the same repository and the same propose route, and
     funding a particular bill is per bill. The rules still ride in the payload -- the projection
     below uses them -- and nothing about funding behaviour changed. */

  const contributions = root.querySelector("[data-payday-contributions]");
  contributions.replaceChildren();
  for (const item of nextRun?.contributions || []) {
    const row = document.createElement("div");
    row.className = "m-payday-contribution";
    const name = document.createElement("span");
    name.textContent = item.commitment;
    const amount = document.createElement("strong");
    amount.textContent = formatCurrency(item.amount, "USD");
    row.append(name, amount);
    contributions.append(row);
  }
  if (!nextRun?.contributions?.length) {
    const empty = document.createElement("p");
    empty.className = "m-settings-empty";
    empty.textContent = "No funding contribution is projected yet.";
    contributions.append(empty);
  }

  renderLearning(payload.learning);
}

async function setLearningFloor(floor) {
  /* The learning window is a Meridian-LOCAL setting: it moves no money, deletes no
     record, and never reaches Crew. So it is posted with plain fetch, the same way the
     owner-initiated connection authorize call is, rather than through the proposal
     channel (which exists for financial proposals) or the Crew mutation helper. */
  const status = root.querySelector("[data-learning-floor-status]");
  const setButton = root.querySelector("[data-learning-floor-set]");
  const clearButton = root.querySelector("[data-learning-floor-clear]");
  setButton.disabled = true;
  clearButton.disabled = true;
  try {
    const response = await fetch("/api/meridian/settings/payday/learning-floor", {
      method: "POST",
      credentials: "same-origin",
      headers: { Accept: "application/json", "Content-Type": "application/json" },
      body: JSON.stringify({ floor }),
    });
    const payload = await response.json().catch(() => null);
    if (!response.ok || !payload) {
      const detail = (payload && payload.error) || {};
      status.textContent = `${detail.message || "The learning window could not be changed."} ${
        detail.recovery_action || ""
      }`.trim();
      return;
    }
    renderLearning(payload.learning);
    status.textContent = floor
      ? "Learning re-based. No records were deleted and nothing was sent to Crew."
      : "Learning from all of your income history again. No records were deleted.";
    // Re-read so the recognised schedule and its counts reflect the new window.
    render(await meridianFetch("/api/meridian/settings/payday"));
    status.textContent = floor
      ? "Learning re-based. No records were deleted and nothing was sent to Crew."
      : "Learning from all of your income history again. No records were deleted.";
  } catch (error) {
    status.textContent = `${error.message} ${error.recoveryAction || ""}`.trim();
  } finally {
    setButton.disabled = false;
    clearButton.disabled = false;
  }
}

function renderFundingPlans(payload) {
  /* Crew's funding plans, so a plan can be ADDRESSED by its id. Names come from a provider, so
     every value renders through textContent -- never innerHTML. */
  const list = root?.querySelector("[data-funding-plans]");
  if (!list) return;
  const plans = Array.isArray(payload?.funding_plans) ? payload.funding_plans : [];
  list.replaceChildren();
  if (!plans.length) {
    const empty = document.createElement("p");
    empty.className = "m-payday-note";
    empty.textContent =
      "Crew is not reporting a funding plan right now, so there is no payday cadence to set. " +
      "Meridian will offer this as soon as Crew returns one.";
    list.append(empty);
    return;
  }
  for (const plan of plans) {
    const row = document.createElement("div");
    row.className = "m-funding-plan-row";

    const identity = document.createElement("div");
    identity.className = "m-funding-plan-identity";
    const name = document.createElement("strong");
    name.textContent = plan.name || "Unnamed plan";
    /* No per-row cadence line any more. The summary card owns the cadence, resolved in the app's
       own precedence and naming the record it came from; two cadence displays from two sources was
       the inconsistency the owner reported (OS-104). This row says which record it is and what it
       pays, which is what its control needs. */
    identity.append(name);

    const input = document.createElement("input");
    input.type = "text";
    input.inputMode = "decimal";
    input.className = "m-funding-plan-amount";
    input.dataset.fundingPlanAmount = "";
    input.value = plan.amount === null || plan.amount === undefined ? "" : String(plan.amount);
    input.setAttribute("aria-label", `Payday amount for ${plan.name || "this plan"}`);

    const button = document.createElement("button");
    button.type = "button";
    button.className = "m-primary-button";
    button.dataset.fundingPlanSet = plan.id || "";
    button.textContent = "Set in Crew";

    button.addEventListener("click", () => setFundingPlan(plan.id, input.value, button));

    row.append(identity, input, button);
    list.append(row);
  }
}

async function setFundingPlan(planId, rawAmount, button) {
  /* A real Crew write, so it goes through the existing action pipeline rather than a bespoke
     route: POST /api/actions/mutate with provenance owner_direct. The ROUTER decides what happens
     -- a direct, fully-specified owner edit executes; anything interpreted, composed or
     low-confidence becomes a proposal -- so this reports whichever actually occurred instead of
     assuming success. Never retried: an uncertain financial mutation must not be repeated. */
  const status = root.querySelector("[data-funding-plan-status]");
  const amount = Number(String(rawAmount ?? "").trim());
  if (!planId) {
    status.textContent = "That plan has no Crew id, so it cannot be set.";
    return;
  }
  if (!Number.isFinite(amount) || amount <= 0) {
    status.textContent = "Enter the payday amount as a number greater than zero.";
    return;
  }
  button.disabled = true;
  try {
    const response = await fetch("/api/actions/mutate", {
      method: "POST",
      credentials: "same-origin",
      headers: { Accept: "application/json", "Content-Type": "application/json" },
      body: JSON.stringify({
        type: "update_crew_paycheck_funding_plan",
        params: { fundingPlanId: planId, amount },
        provenance: "owner_direct",
        rationale: "owner set the payday amount directly in Settings",
      }),
    });
    const payload = await response.json().catch(() => null);
    if (!response.ok || !payload) {
      const detail = (payload && payload.error) || {};
      status.textContent = `${
        detail.message || "The payday amount could not be set."
      } ${detail.recovery_action || ""}`.trim();
      return;
    }
    if (payload.routing_direct) {
      status.textContent =
        "Set in Crew. Meridian reads Crew back before treating it as confirmed, so this may show " +
        "as pending until that read completes.";
    } else {
      status.textContent =
        "That change needs your approval, so it is waiting in Pending Actions. Nothing has " +
        "reached Crew yet.";
    }
    renderFundingPlans(await meridianFetch("/api/meridian/settings/payday"));
  } catch (error) {
    status.textContent = `${error.message} ${error.recoveryAction || ""}`.trim();
  } finally {
    button.disabled = false;
  }
}

async function load() {
  if (!root) return;
  try {
    const payload = await meridianFetch("/api/meridian/settings/payday");
    render(payload);
    renderFundingPlans(payload);
  } catch (error) {
    const errorBox = root.querySelector("[data-payday-error]");
    errorBox.textContent = `${error.message} ${error.recoveryAction || ""}`.trim();
    errorBox.hidden = false;
  }
}

root?.querySelector("[data-learning-floor-set]")?.addEventListener("click", () => {
  const input = root.querySelector("[data-learning-floor-input]");
  const value = input?.value || "";
  if (!value) {
    root.querySelector("[data-learning-floor-status]").textContent =
      "Choose the date your new pay history starts.";
    return;
  }
  setLearningFloor(value);
});
root?.querySelector("[data-learning-floor-clear]")?.addEventListener("click", () => {
  setLearningFloor(null);
});
load();

File: meridian/services/payday.py
"""Read-only Payday & Funding settings composed from existing Meridian data."""

from __future__ import annotations

from datetime import date
from decimal import Decimal

from meridian.funding import project_funding
from meridian.paycheck_learning import (
    PaycheckLearningFloorRepository,
    observations_on_or_after,
)
from meridian.payday import recognize_payday
from meridian.services.today import data_freshness


def _rule_payload(rule, commitment) -> dict[str, object]:
    return {
        "id": rule.id,
        "commitment_id": rule.commitment_id,
        "commitment": commitment.name if commitment is not None else "Unknown commitment",
        "kind": rule.kind,
        "amount": float(rule.amount) if rule.amount is not None else None,
        "percent": float(rule.percent) if rule.percent is not None else None,
        "cadence": rule.cadence,
        "paused": rule.paused,
        "priority": rule.priority,
    }


def build_learning_window(graph, transactions=None) -> dict[str, object]:
    """The owner's learnable window, as a reportable payload (OS-051).

    One builder, used by both the settings read and the reset endpoint, so the control
    and the report can never disagree about what the floor currently excludes. The
    excluded count is the point of it: a reset whose effect is invisible is not
    governable.
    """
    stored = PaycheckLearningFloorRepository(graph.db_path).get()
    floor = stored.floor if stored is not None else None
    if transactions is None:
        transactions, _cursor = graph.list_transactions(limit=200)
    included = len(observations_on_or_after(transactions, floor))
    return {
        "active": floor is not None,
        "floor": floor,
        "set_at": stored.set_at if stored is not None else None,
        "included": included,
        "excluded": len(transactions) - included,
    }


def resolve_cadence(pattern, funding_plans) -> dict[str, object]:
    """THE cadence, resolved in the order the rest of the app already uses.

    `meridian/api.py::_paycheck_config` has answered this question the same way since 2026-09-19: Crew's
    funding plan outranks every other leg, on the owner's directive that the paycheck "SHOULD be a crew
    record", and `meridian/providers/base.py::FundingPlanCandidate` calls that record the owner's "income
    source / Funding Cadence". This settings payload reported only the OBSERVED pattern instead, so the
    Cadence card could read "Not recognized" beside a Crew cadence listed lower on the same page
    (owner-reported 2026-09-24: "payday is the funding mechanism, so the payday and when it lands is the
    cadence, this isn't consistent").

    One resolver, one answer, and the source is named so the card can say where the figure came from
    rather than implying the app does not know. When Crew reports more than one distinct cadence the
    card names the record it used and says how many others exist -- a single figure must not silently
    stand in for several.
    """
    crew = [
        plan
        for plan in funding_plans
        if isinstance(plan, dict) and plan.get("cadence")
    ]
    if crew:
        governing = crew[0]
        distinct = {str(plan["cadence"]) for plan in crew}
        return {
            "value": governing["cadence"],
            "source": "crew",
            "source_label": "Crew paycheck",
            "plan_id": governing.get("id"),
            "plan_name": governing.get("name"),
            "other_records": len(crew) - 1,
            "distinct_cadences": len(distinct),
            "conflict": len(distinct) > 1,
        }
    if pattern is not None:
        return {
            "value": pattern.cadence,
            "source": "observed",
            "source_label": "Learned from your deposits",
            "confidence": pattern.confidence,
            "deposits": len(pattern.evidence_ids),
            "other_records": 0,
            "distinct_cadences": 1,
            "conflict": False,
        }
    return {
        "value": None,
        "source": None,
        "source_label": None,
        "other_records": 0,
        "distinct_cadences": 0,
        "conflict": False,
    }


def build_payday_settings(graph, commitments, rules, *, as_of: date) -> dict[str, object]:
    transactions, _cursor = graph.list_transactions(limit=200)

    # Crew's funding plans, so the surface can ADDRESS one. The plan id is Crew's own
    # external_id and is the identity the write path needs: the write operations are
    # update/delete_crew_paycheck_funding_plan, and the manifest records that the update is
    # 'Identified by the approved proposal's fundingPlanId'. The NAME is deliberately not the
    # key -- the record's own docstring says the owner renames the plan, so anything keyed on
    # the name breaks on the next rename. Best-effort like the paycheck context path: a read
    # failure must not take the whole settings page down.
    funding_plans: list[dict[str, object]] = []
    try:
        from meridian.repository import FinancialRepository

        financial = (
            graph
            if isinstance(graph, FinancialRepository)
            else FinancialRepository(graph.db_path)
        )
        funding_plans = [
            {
                "id": record.external_id,
                "name": record.name,
                "amount": record.amount,
                "cadence": record.cadence,
            }
            for record in financial.list_funding_plans()
            if getattr(record, "absent_since", None) is None
        ]
    except Exception:  # noqa: BLE001 - observations are best-effort context
        funding_plans = []
    # The owner's learning window (OS-051). It governs which observations may be learned
    # from, so it is read BEFORE recognition and reported alongside the result.
    learning = build_learning_window(graph, transactions)
    pattern = recognize_payday(transactions, as_of=as_of, floor=learning["floor"])
    rule_views = []
    next_contributions = []
    for rule in rules.list_all():
        commitment = commitments.get(int(rule.commitment_id))
        rule_views.append(_rule_payload(rule, commitment))
        if pattern is None or commitment is None:
            continue
        projection = project_funding(
            rule,
            commitment,
            [(pattern.next_date, Decimal(str(pattern.typical_amount)))],
            as_of=as_of,
        )
        for event in projection.events:
            if event.date == pattern.next_date and event.amount > 0:
                next_contributions.append(
                    {
                        "rule_id": rule.id,
                        "commitment": commitment.name,
                        "amount": float(event.amount),
                        "explanation": list(event.explanation),
                    }
                )

    cash_accounts = [
        account
        for account in graph.list_accounts()
        if account.is_active and account.account_type in {"cash", "checking", "savings"}
    ]
    next_run = None
    if pattern is not None:
        next_run = {
            "date": pattern.next_date.isoformat(),
            "kind": "proposal",
            "total": round(sum(item["amount"] for item in next_contributions), 2),
            "contributions": next_contributions,
        }
    return {
        "state": "current" if pattern is not None else "unavailable",
        # ONE cadence for the surface, resolved in the app's own precedence rather than from the
        # observed pattern alone (OS-104). The observed pattern is still reported separately below,
        # because "what Crew says" and "what Meridian has seen" are different facts and both matter.
        "cadence": resolve_cadence(pattern, funding_plans),
        "pattern": (
            {
                "cadence": pattern.cadence,
                "next_date": pattern.next_date.isoformat(),
                "typical_amount": pattern.typical_amount,
                "confidence": pattern.confidence,
                "evidence_count": len(pattern.evidence_ids),
            }
            if pattern is not None
            else None
        ),
        "funding_source": (
            {
                "account_id": cash_accounts[0].id,
                "name": cash_accounts[0].name,
            }
            if cash_accounts
            else None
        ),
        "funding_plans": funding_plans,
        "rules": rule_views,
        "next_run": next_run,
        "learning": learning,
        "data_freshness": data_freshness(graph, include_all_connections=True),
    }
